[WIP] Improve Snaps concepts and API reference (#1098)
MetaMask/metamask-docs: snaps/reference/permissions.md

sidebar_position: 4
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Snaps permissions

Snaps can [request the following permissions](../how-to/request-permissions.md).

## RPC API permissions

You must request permission to call [Snaps API methods](snaps-api.md).

> For example, to request to call [`snap_dialog`](snaps-api.md#snap_dialog), add the following to the

Reviewer: ```suggestion
For example, to call the [`snap_dialog`](snaps-api.md#snap_dialog) function, add the following to the
```

Author: If you read the sentence before this, it really is "requesting to call" an API method. And to be consistent across the docs, we should refer to API methods as "methods" rather than "functions." In most of our docs, it's also okay to refer directly to the method names without saying "the X method" (in this case, we mention "methods" directly in the previous sentence), so I'll leave this as is.

Specify this permission in the manifest file as follows:

```json title="snap.manifest.json"
"initialPermissions": {
  "endowment:cronjob": {
    "jobs": [
      {
        "expression": "* * * * *",
        "request": {
          "method": "exampleMethodOne",
          "params": {
            "param1": "foo"
          }
        }
      },
      {
        "expression": "*/2 * * * *",
        "request": {
          "method": "exampleMethodTwo",
          "params": {
            "param1": "bar"
          }
        }
      }
    ]
  }
}

```

### endowment:ethereum-provider

To communicate with a node using MetaMask, a Snap must request the `endowment:ethereum-provider` permission.
This permission exposes the `ethereum` global to the Snap execution environment, allowing Snaps to
call some [MetaMask JSON-RPC API](/wallet/reference/json-rpc-api) methods.

> This global is an [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193) provider.

Reviewer: Sorry, is it a global object?

Specify this permission in the manifest file as follows:

```json title="snap.manifest.json"
"initialPermissions": {
  "endowment:ethereum-provider": {}
}
```

:::note 

> The `ethereum` global available to Snaps has fewer capabilities than `window.ethereum` for dapps. 

Reviewer: ```suggestion
The global `ethereum` object in Snaps has fewer capabilities than `window.ethereum` for dapps.
```

Specify this permission in the manifest file as follows:

```json title="snap.manifest.json"
"initialPermissions": {
  "endowment:webassembly": {}
}
```

## Dynamic permissions

### eth_accounts

A Snap can request permission to call the [`eth_accounts`](/wallet/reference/eth_accounts) MetaMask
JSON-RPC API method by calling [`eth_requestAccounts`](/wallet/reference/eth_requestaccounts).
Calling `eth_requestAccounts` requires the
[`endowment:ethereum-provider`](#endowmentethereum-provider) permission:

<Tabs>
<TabItem value="Manifest file">

```json
"initialPermissions": {
  "endowment:ethereum-provider": {}
}
```

</TabItem>
<TabItem value="JavaScript">

```js
await ethereum.request({
  "method": "eth_requestAccounts"
});
```

</TabItem>
</Tabs>

You can check the presence of the permission by calling
[`wallet_getPermissions`](/wallet/reference/wallet_getpermissions).
If the permission is present, the result contains a permission with a `parentCapability` of `eth_accounts`.

> It comes with a caveat of `restrictReturnedAccounts`, an array of all the accounts the user allows

Author: "Caveats" is a technical term here - right below this line is an example of what this describes. I'll update the sentence to read "The permission contains a `restrictReturnedAccounts` caveat, ..."
